src/meta-srv/src/handler/failure_handler/runner.rs

```diff
@@ -0,0 +1,300 @@
+        let failure_detectors = Box::new(DefaultFailureDetectorContainer(HashMap::new()));
+        self.start_with(failure_detectors).await
+    }
+
+    async fn start_with(&mut self, failure_detectors: Box<dyn FailureDetectorContainer>) {
+        let Some(mut heartbeat_rx) = self.heartbeat_rx.take() else { return };
+        let Some(mut control_rx) = self.control_rx.take() else { return };
+
+        let election = self.election.clone();
+        let mut failure_detectors = failure_detectors;
+        let handle = common_runtime::spawn_bg(async move {
+            loop {
+                loop {
+                    let maybe_control = control_rx.try_recv();
+                    match maybe_control {
+                        Ok(control) => match control {
+                            FailureDetectControl::Purge => failure_detectors.clear(),
+
+                            #[cfg(test)]
+                            FailureDetectControl::Dump(tx) => {
+                                // Drain any heartbeats that are not handled before dump.
+                                while let Ok(heartbeat) = heartbeat_rx.try_recv() {
+                                    for ident in heartbeat.region_idents {
+                                        let detector =
+                                            failure_detectors.get_failure_detector(ident);
+                                        detector.heartbeat(heartbeat.heartbeat_time);
+                                    }
+                                }
+                                let _ = tx.send(failure_detectors.dump());
+                            }
+                        },
+                        Err(TryRecvError::Empty) => break,
+                        Err(TryRecvError::Disconnected) => return,
+                    }
+                }
+
+                loop {
+                    let maybe_heartbeat = heartbeat_rx.try_recv();
+                    match maybe_heartbeat {
+                        Ok(heartbeat) => {
+                            for ident in heartbeat.region_idents {
+                                let detector = failure_detectors.get_failure_detector(ident);
+                                detector.heartbeat(heartbeat.heartbeat_time);
+                            }
+                        }
+                        Err(TryRecvError::Empty) => break,
+                        Err(TryRecvError::Disconnected) => return,
+                    }
+                }
+
+                let is_leader = election.as_ref().map(|x| x.is_leader()).unwrap_or(true);
+                if is_leader {
+                    let now = current_time_millis();
+                    for detector in failure_detectors.iter() {
+                        if !detector.is_available(now) {
+                            // TODO(LFC): TBC
+                        }
+                    }
+                }
+
+                tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
Sleep here for a second, I am worried that a large amount of heartbeat data will be accumulated in the channel.